### PR TITLE
[CE] Reduce memory consumption of queued messages

### PR DESCRIPTION
**Pull request by DC2-DanielKrueger**
**Motivation**

Resolves #<issue>

**Changes**

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -150,6 +150,9 @@ dependencies {
     implementation("net.openhft:zero-allocation-hashing:${property("zero-allocation-hashing.version")}")
     implementation("com.fasterxml.jackson.core:jackson-databind:${property("jackson.version")}")
     implementation("org.jctools:jctools-core:${property("jctools.version")}")
+
+    /* primitive data structures */
+    implementation("org.eclipse.collections:eclipse-collections:10.4.0")
```
Comment (SgtSilvio) [inline build.gradle.kts:155]: please use a version property (as with all other dependencies)

File: src/main/java/com/hivemq/persistence/payload/BucketLock.java [hunk 2]
```diff
@@ -15,26 +15,54 @@
...
+import static com.google.common.base.Preconditions.checkNotNull;
+
 /**
  * @author Lukas Brandl
  */
 public class BucketLock {
 
-    private final Lock[] locks;
+    private final Lock @NotNull [] locks;
+    private final int bucketCount;
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/persistence/payload/BucketLock.java:32]: remove, use locks.length instead

File: src/main/java/com/hivemq/persistence/payload/PublishPayloadPersistenceImpl.java [hunk 3]
```diff
@@ -59,10 +55,13 @@
     private final @NotNull BucketLock bucketLock;
 
     @NotNull Cache<Long, byte[]> payloadCache;
-    final ConcurrentHashMap<Long, AtomicLong> referenceCounter = new ConcurrentHashMap<>();
     final Queue<RemovablePayload> removablePayloads = new LinkedTransferQueue<>();
+    private final @NotNull PayloadReferenceCounterRegistry payloadReferenceCounterRegistry;
+
 
     private @Nullable ListenableScheduledFuture<?> removeTaskFuture;
+    private final int bucketLockCount;
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/persistence/payload/PublishPayloadPersistenceImpl.java:63]: remove, use `bucketLock.getBucketCount()` instead

File: src/test/java/com/hivemq/persistence/payload/PublishPayloadPersistenceImplTest.java [hunk 4]
```diff
@@ -143,44 +143,44 @@ public void get_from_local_persistence_retained_message_null_payload() throws Ex
...
         assertEquals(1, persistence.removablePayloads.size());
     }
 
     @Test
     public void decrement_reference_count_already_zero() throws Exception {
-        persistence.referenceCounter.put(0L, new AtomicLong(0L));
+        persistence.getPayloadReferenceCounterRegistry().put(0L, 0);
         persistence.decrementReferenceCounter(0L);
-        assertEquals(0L, persistence.referenceCounter.get(0L).get());
+        assertEquals(0L, (long)persistence.getReferenceCountersAsMap().get(0L));
```
Comment (SgtSilvio) [inline src/test/java/com/hivemq/persistence/payload/PublishPayloadPersistenceImplTest.java:176]:
```suggestion
        assertEquals(0, (int) persistence.getReferenceCountersAsMap().get(0));
```

File: src/test/java/com/hivemq/persistence/payload/PublishPayloadPersistenceImplTest.java [hunk 5]
```diff
@@ -143,44 +143,44 @@ public void get_from_local_persistence_retained_message_null_payload() throws Ex
...
         assertEquals(0, persistence.removablePayloads.size());
     }
 
     @Test
     public void decrement_reference_count_to_zero() throws Exception {
-        persistence.referenceCounter.put(0L, new AtomicLong(1L));
+        persistence.getPayloadReferenceCounterRegistry().put(0L, 1);
         persistence.decrementReferenceCounter(0L);
-        assertEquals(0L, persistence.referenceCounter.get(0L).get());
+        assertEquals(0L, (long)persistence.getReferenceCountersAsMap().get(0L));
```
Comment (SgtSilvio) [inline src/test/java/com/hivemq/persistence/payload/PublishPayloadPersistenceImplTest.java:168]:
```suggestion
        assertEquals(0, (int) persistence.getReferenceCountersAsMap().get(0));
```

File: src/test/java/com/hivemq/persistence/payload/PublishPayloadPersistenceImplTest.java [hunk 6]
```diff
@@ -143,44 +143,44 @@ public void get_from_local_persistence_retained_message_null_payload() throws Ex
...
+        assertEquals(2L, (long)persistence.getReferenceCountersAsMap().get(0L));
     }
 
     @Test
     public void decrement_reference_count() throws Exception {
-        persistence.referenceCounter.put(0L, new AtomicLong(2L));
+        persistence.getPayloadReferenceCounterRegistry().put(0L,2);
         persistence.decrementReferenceCounter(0L);
-        assertEquals(1L, persistence.referenceCounter.get(0L).get());
+        assertEquals(1L, (long)persistence.getReferenceCountersAsMap().get(0L));
```
Comment (SgtSilvio) [inline src/test/java/com/hivemq/persistence/payload/PublishPayloadPersistenceImplTest.java:160]:
```suggestion
        assertEquals(1, (int) persistence.getReferenceCountersAsMap().get(0));
```

File: src/main/java/com/hivemq/persistence/payload/BucketLock.java [hunk 7]
```diff
@@ -15,26 +15,54 @@
...
 
-    public Lock get(final String key) {
-        return locks[BucketUtils.getBucket(key, locks.length)];
+
+    public int getBucketCount() {
+        return bucketCount;
+    }
+
+
+    public void accessBucketByPaloadId(final @NotNull long payloadId, final @NotNull BucketAccessCallback callback) {
```
Comment (Remit) [inline src/main/java/com/hivemq/persistence/payload/BucketLock.java:48]:
```suggestion
    public void accessBucketByPayloadId(final @NotNull long payloadId, final @NotNull BucketAccessCallback callback) {
```

File: src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistry.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.Map;
+
+/**
+ * @author Daniel Krüger
+ */
+public interface PayloadReferenceCounterRegistry {
+
+
+    static String toPayloadId(final long nodeID, final long counter) {
+        return nodeID + "_pub_" + counter;
```
Comment (Remit) [inline src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistry.java:30]: _pub_ to constant maybe?

File: src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistry.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import com.hivemq.extension.sdk.api.annotations.Nullable;
+
+import java.util.Map;
+
+/**
+ * @author Daniel Krüger
+ */
+public interface PayloadReferenceCounterRegistry {
```
Comment (Remit) [inline src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistry.java:26]: javadoc would be nice, especially in CE

File: src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistry.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+    int decrement(final long payloadId);
+
+    @NotNull Map<Long, Integer> getAll();
+
+    int size();
+
+    void remove(@NotNull long payloadId);
+
+
```
Comment (Remit) [inline src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistry.java:51]:
```suggestion
```

File: src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistryImpl.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+/**
+ * @author Daniel Krüger
+ * <p>
+ * Most methods are NOT thread-safe and the exclusive access on the bucket must be secured by the caller
+ * The reason is that the caller (primarly PublishPayloadPersistence) calls often multiple methods and the lock must cover all sequential calls to methods
+ */
+public class PayloadReferenceCounterRegistryImpl implements PayloadReferenceCounterRegistry {
+
+    private final int numberBuckets;
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistryImpl.java:39]: field can be replaced by buckets.length

File: src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistryImpl.java [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+        this.numberBuckets = bucketLock.getBucketCount();
+        this.bucketLock = bucketLock;
+        this.buckets = new LongIntHashMap[numberBuckets];
+        for (int i = 0; i < numberBuckets; i++) {
+            this.buckets[i] = new LongIntHashMap();
+        }
+    }
+
+    @NotThreadSafe
+    public @Nullable int get(final long payloadId) {
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistryImpl.java:53]:
```suggestion
    public int get(final long payloadId) {
```

File: src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistryImpl.java [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    PayloadReferenceCounterRegistryImpl(final @NotNull BucketLock bucketLock) {
+        this.numberBuckets = bucketLock.getBucketCount();
+        this.bucketLock = bucketLock;
+        this.buckets = new LongIntHashMap[numberBuckets];
+        for (int i = 0; i < numberBuckets; i++) {
+            this.buckets[i] = new LongIntHashMap();
+        }
+    }
+
+    @NotThreadSafe
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistryImpl.java:52]: please annotate the class as threadsafe, override missing

File: src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistryImpl.java [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+            return;
+        }
+        map.remove(payloadId);
+    }
+
+    @ThreadSafe
+    public @NotNull ImmutableMap<Long, Integer> getAll() {
+        final ImmutableMap.Builder<Long, Integer> builder = ImmutableMap.builder();
+        for (int i = 0; i < numberBuckets; i++) {
+            final int finalI = i;
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistryImpl.java:99]: finalI -> bucketIndex

File: src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistryImpl.java [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+            });
+        }
+        return builder.build();
+    }
+
+    @ThreadSafe
+    public int size() {
+        final AtomicInteger sum = new AtomicInteger();
+        for (int i = 0; i < numberBuckets; i++) {
+            final int finalI = i;
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/persistence/payload/PayloadReferenceCounterRegistryImpl.java:113]: finalI -> bucketIndex

File: src/main/java/com/hivemq/persistence/payload/PublishPayloadPersistenceImpl.java [hunk 16]
```diff
@@ -94,34 +95,24 @@ public void init() {
...
         }
     }
 
-    /**
-     * {@inheritDoc}
-     */
-    @Override
-    public boolean add(@NotNull final byte[] payload, final long referenceCount, final long payloadId) {
+
+    public boolean add(final byte @NotNull [] payload, final long referenceCount, final @NotNull long payloadId) {
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/persistence/payload/PublishPayloadPersistenceImpl.java:104]:
```suggestion
    public boolean add(final byte @NotNull [] payload, final long referenceCount, final long payloadId) {
```

File: src/main/java/com/hivemq/persistence/payload/PublishPayloadPersistenceImpl.java [hunk 17]
```diff
@@ -188,67 +174,44 @@ public void incrementReferenceCounterOnBootstrap(final long payloadId) {
...
-        }
-
+        });
     }
 
-
     @Override
     public void closeDB() {
         if (removeTaskFuture != null) {
-            removeTaskFuture.cancel(true);
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/persistence/payload/PublishPayloadPersistenceImpl.java:223]: ?

File: src/main/java/com/hivemq/persistence/payload/RemoveEntryTask.java [hunk 18]
```diff
@@ -63,27 +64,18 @@ public void run() {
...
-                            //The reference count can be null, if it was marked as removable twice.
-                            //Which is possible if a payload marked as removable and we receive the same payload again and mark it as removable again,
-                            //before the cleanup is able to remove the payload.
-                            removablePayload = removablePayloads.poll();
-                            continue;
-                        }
-                        if (referenceCount.get() == 0) {
+                    bucketLock.accessBucketByPaloadId(removablePayload.getId(), ()->{
+                        final int referenceCount = payloadReferenceCounterRegistry.get(payloadId);
+                        //The reference count can be null, if it was marked as removable twice.
```
Comment (SgtSilvio) [inline src/main/java/com/hivemq/persistence/payload/RemoveEntryTask.java:70]: null -> UNKNOWN_PAYLOAD

File: src/test/java/com/hivemq/persistence/payload/PublishPayloadPersistenceImplTest.java [hunk 19]
```diff
@@ -76,8 +76,8 @@ public void add_new_entries() throws Exception {
         persistence.add(payload1, 1, 123);
         persistence.add(payload2, 2, 234);
 
-        assertEquals(1, persistence.referenceCounter.get(123L).get());
-        assertEquals(2, persistence.referenceCounter.get(234L).get());
+        assertEquals(1, (long)persistence.getReferenceCountersAsMap().get(123L));
```
Comment (SgtSilvio) [inline src/test/java/com/hivemq/persistence/payload/PublishPayloadPersistenceImplTest.java:79]: the long casts should be unnecessary

**Review by SgtSilvio — APPROVED** [review]

**Pull request merged**